lib/mongodb.js

```diff
@@ -1881,14 +1881,30 @@ MongoDB.prototype.isObjectIDProperty = function(model, prop, value, options) {
   ) {
     return true;
   } else if ('string' === typeof value) {
+    // dataType is specified as ObjectID
+    if (prop && prop.mongodb && prop.mongodb.dataType === ObjectID) {
+      // value looks like an ObjectID, and it should be treated as one
+      // because `prop.mongodb.dataType` is set to 'ObjectID'
+      // this condition overrides the `strictObjectIDCoercion` setting
+      if (value.match(/^[0-9a-fA-F]{24}$/)) return true;
```

Comment:
This regular expression is repeated in quite few places, please extract it into a shared constant.